### PR TITLE
Smallest step to get a streamed request

### PR DESCRIPTION
**Pull request by davidbgk**
An extract from https://github.com/pyrates/roll/pull/77/ without all the hooks.

I *really* don't know what I'm doing, tests are running fine though. If you think we can remove even more stuff, be my guest. The aim of that PR is to make a baby step toward being fully async.

### CODE REVIEW

File: tests/test_request.py [hunk 1]
```diff
@@ -520,3 +543,51 @@ def protocol(app, event_loop):
...
+async def test_can_consume_lazy_request_iterating(client, app):
+
+    @app.route('/test', methods=['POST'], lazy=True)
+    async def post(req, resp):
+        async for chunk in req:
+            resp.body = chunk
+
+    resp = await client.post('/test', data='blah'*1000)
+    assert resp.status == HTTPStatus.OK
+    assert resp.body == b'blah'*1000
```
Comment (yohanboniface) [inline tests/test_request.py]: Ah, liveclient tests seem to be missing ?

File: roll/io.py [hunk 2]
```diff
@@ -90,10 +138,12 @@ def files(self):
 
     @property
     def json(self):
-        try:
-            return json.loads(self.body.decode())
-        except (UnicodeDecodeError, JSONDecodeError):
-            raise HttpError(HTTPStatus.BAD_REQUEST, 'Unparsable JSON body')
+        if self._json is None:
```
[the remark this answers is not in the document]
Comment (davidbgk) [reply roll/io.py]: Done in https://github.com/pyrates/roll/pull/90

File: roll/io.py [hunk 3]
```diff
@@ -103,13 +153,42 @@ def content_type(self):
...
+        if self._body is None:
+            raise HttpError(HTTPStatus.INTERNAL_SERVER_ERROR,
+                            "Trying to consume lazy body")
+        return self._body
+
+    @body.setter
+    def body(self, data):
+        self._body = data
+
+    async def read(self):
```
[the remark this answers is not in the document]
Comment (davidbgk) [reply roll/io.py]: Maybe `req.consume()` would be more explicit? Or `load_body()`?

File: roll/io.py [hunk 4]
```diff
@@ -103,13 +153,42 @@ def content_type(self):
...
+        if not self._body:
+            self._body = b''
+            async for chunk in self:
+                self._body += chunk
+        return self._body
+
+    async def __aiter__(self):
+        # TODO raise if already consumed?
```
Comment (davidbgk) [inline roll/io.py]: Or return the whole body? 🤔

File: roll/io.py [hunk 5]
```diff
@@ -103,13 +153,42 @@ def content_type(self):
...
 
 class Response:
     """A container for `status`, `headers` and `body`."""
-    __slots__ = ('app', '_status', 'headers', 'body', '_cookies')
+    __slots__ = ('app', '_status', 'headers', 'body', '_cookies', 'protocol')
 
-    def __init__(self, app):
+    def __init__(self, app, protocol):
         self.app = app
+        self.protocol = protocol
```
[the remark this answers is not in the document]
Comment (davidbgk) [reply roll/io.py]: Agreed, and I'm not sure why we actually need it for the Response, is that to be symmetric with the Request?

File: tests/test_request.py (hunk 1 above)
Comment (davidbgk) [reply tests/test_request.py]: I don't get why that test is not passing anymore, I need to dig into that.

Comment (davidbgk) [reply tests/test_request.py]: I pushed the test (not passing).

Comment (davidbgk) [reply tests/test_request.py]: Wait, it is passing on the CI 🤔

File: roll/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidbgk) [reply roll/io.py]: I renamed the method for `load_body`.

File: tests/test_request.py (hunk 1 above)
Comment (davidbgk) [reply tests/test_request.py]: It doesn't look to pass under python 3.6 with default macOS python, updating to breaw's python3.7 made the test passing.

**Pull request merged**